### PR TITLE
added support for coarse reverse geocoding

### PR DESCRIPTION
**Pull request by trescube**
several things are in this commit:

- coarse reverse controller / pip service
- inject "conditional execute" predicate into search service to shortcut execution conditions
- added coarse reverse controller to precede standard reverse controller
- lots of tests!

### CODE REVIEW

File: controller/coarse_reverse.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
+const logger = require('pelias-logger').get('coarse_reverse');
```
Comment (dianashk) [inline controller/coarse_reverse.js]: Doesn't this need 'use strict'?

File: controller/predicates/is_coarse_reverse.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+const _ = require('lodash');
+
+module.exports = (req, res) => {
+  // returns true if layers is undefined, empty, or contains 'address' or 'venue'
+  return !_.isEmpty(req.clean.layers) &&
+          _.intersection(req.clean.layers, ['address', 'venue']).length === 0;
```
Comment (dianashk) [inline controller/predicates/is_coarse_reverse.js]: Should street layer be in this list?

File: controller/predicates/is_pip_service_enabled.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const _ = require('lodash');
+
+module.exports = (uri) => {
+  // this predicate relies upon the fact that the schema has already validated
+  //  that api.pipService is a URI-formatted string
+  return (request, response) => {
+    return uri !== undefined;
+  };
```
Comment (dianashk) [inline controller/predicates/is_pip_service_enabled.js]: Use the lodash isSet function or remove lodash from the top

Comment (dianashk) [inline controller/predicates/is_pip_service_enabled.js]: isSet is the wrong function, sorry. It checks that the value is of type Set, which I always forget! but there is _.isUndefined()

File: controller/coarse_reverse.js (hunk 1 above)
Comment (trescube) [inline controller/coarse_reverse.js]: `const` doesn't, `let` does.

File: routes/v1.js [hunk 4]
```diff
@@ -80,9 +111,9 @@ function addRoutes(app, peliasConfig) {
       middleware.calcSize(),
       // 3rd parameter is which query module to use, use fallback/geodisambiguation
       //  first, then use original search strategy if first query didn't return anything
-      controllers.search(peliasConfig.api, esclient, queries.libpostal),
+      controllers.search(peliasConfig.api, esclient, queries.libpostal, not(hasDataOrErrors)),
```
Comment (trescube) [inline routes/v1.js]: Good point, I'll fix it.

File: routes/v1.js [hunk 5]
```diff
@@ -1,6 +1,10 @@
 var Router = require('express').Router;
 var elasticsearch = require('elasticsearch');
 
+const all = require('predicates').all;
+const any = require('predicates').any;
+const not = require('predicates').not;
```
Comment (missinglink) [inline routes/v1.js]: yea this is cool

**Pull request merged**